### ISSUE TITLE
Proposal: Add ability to have screenshots masked

### ISSUE THREAD

**Issue by SetTrend** [w3c/webdriver#1286]
Screenshots containing real-time data can not be compared to screenshots taken in a previous test session. So, it is not possible for automated test runs to tell changes in layout from changes in data.

In order to validate screenshots in automated web tests it is necessary to be able to exclude real-time data from screenshots supposed to be validated.
<br/>

<hr/>

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Original Screenshot:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Masking Realtime Data:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Resulting Masked Image:
<img alt="Original Screenshot" src="https://user-images.githubusercontent.com/9283914/44159905-16feee80-a0b9-11e8-8cbe-8b8d1f854788.png" width="30%"/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img alt="Masking Realtime Datea" src="https://user-images.githubusercontent.com/9283914/44157511-ecaa3280-a0b2-11e8-9b05-8ede38ee01bd.png" width="30%"/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img alt="Resulting Masked Image" src="https://user-images.githubusercontent.com/9283914/44159924-23834700-a0b9-11e8-9dbb-442a1d97e4fe.png" width="30%"/>

<hr>
<br/>

The rightmost (masked) image in the above example will always be the same across test runs. Hence, an automated test may be able to compare a masked image taken at a previous test run to a screenshot taken at a current test run.

One may argue that such masks could be manually created, but this would contradict the purpose of automated testing. Moreover, it would result in extraordinary loads of redundant manpower for manually creating masks in cross-device tests. This is particularly true for volatile test pages created in agile development environments.
<br/>
<hr>

#### For this purpose two things are required:

1. Some kind of tag that's telling WebDriver to mask a particular inline/block element in a document.

1. A capability to give the mask color (in order for test validation algorithms to automatically identify a mask in an image if that's a requirement)

#### Just to give a start of thought, I'd like to propose this kind of implementation:

A new WebDriver capability may be introduced, called "`imageMask`". It is supposed to be an Object with the following TypeScript definition:

```ts
interface imageMask
{ public selector: string
, public color: string
}
```

1. The "`imageMask.selector`" property shall be accepting any [CSS3 selector](https://www.w3.org/TR/selectors/#overview), telling a WebDriver implementation which document elements to look out for when it's supposed to render masks in a screenshot.

1. The "`imageMask.color`" property shall be accepting any [CSS3 color value](https://www.w3.org/TR/css-color-3/#colorunits), having the alpha channel usually being set to "`1`", but allowing for semi-transparent masks, too. The WebDriver implementation is supposed to overlay regions to be masked with this color.

#### Given my above proposal, the suggested WebDriver capability may look like this:
<details>

```json
{ "imageMask":
  { "selector": "div[wd-mask]"
  , "color": "rgba(255,0,255,1)"
  }
}
```
... or, like this:
```json
{ "imageMask":
  { "selector": "div[wd-mask]"
  , "color": "FF00FF"
  }
}
```
... or, like this:
```json
{ "imageMask":
  { "selector": "div[wd-mask]"
  , "color": "hsl(300,100,50)"
  }
}
```
</details>

#### In this example, the above web page's HTML source may look like this:
<details>

```html
<!doctype html>
<html>

<head>
  <title>Fast Brokers</title>
</head>

<body>
  <header>
    <h1>Fast Brokers</h1>
    <div>
      <div>
        <a href="#"><img alt="Login" src="..."/></a>
        <a href="#"><img alt="Search" src="..."/></a>
      </div>
      <div wd-mask="wd-mask">02/05/2019&ensp;10:52 a.m.</div>
    </div>
  </header>

  <main>
    <h2>Twyx Corp</h2>

    <table>
      <tbody>
        <tr><th>WKN:</th><td>A0M2QP</td></tr>
        <tr><th>ISIN:</th><td>CNE1020634Y4</td></tr>
        <tr><th>Sector:</th><td>Telecommunication</td></tr>
        <tr><th>Country:</th><td>China</td></tr>
      </tbody>
    </table>

      <ul>
        <li><a href="#">Overview</a></li>
        <li><a href="#">Charts</a></li>
        <li><a href="#">News</a></li>
      </ul>

      <div wd-mask="wd-mask">$ 105.75</div>
      <div wd-mask="wd-mask">+ $ 2.11&ensp;/&ensp;7.24 %</div>
      <div wd-mask="wd-mask">02/05/2019 10:45 a.m., direct trade </div>

      <ul>
        <li><a href="#">Watch</a></li>
        <li><a href="#">Buy</a></li>
        <li><a href="#">Sell</a></li>
      </ul>
  </main>

  <footer>
    <a href="#">contact us</a> |
    <a href="#">about</a> |
    <a href="#">feedback</a> |
    <a href="#">forums</a> |
  </footer>
</body>

</html>
```
</details>

#### The excerpt from the above HTML document, containing the important parts only:
<br/>

```html
      <div wd-mask="wd-mask">02/05/2019&ensp;10:52 a.m.</div>
```
&hellip;
```html
      <div wd-mask="wd-mask">$ 105.75</div>
      <div wd-mask="wd-mask">+ $ 2.11&ensp;/&ensp;7.24 %</div>
      <div wd-mask="wd-mask">02/05/2019 10:45 a.m., direct trade </div>
```

**Comment by SetTrend** [comment]
Can you please elaborate on your downvotes? What's irritating you?

**Comment by marioklump** [comment]
> Can you please elaborate on your downvotes? What is irritating you?

I'm also interested in the reason for the downvotes. We currently also have the challenge to exclude some areas from comparing previous and current screenshots. It there another solution for excluding some areas so that this proposal is deprecated?

**Comment by ozcoder** [comment]
I like the idea, but I'm guessing the down votes by other people are because it's not really a **webdriver** issue. It's a image comparison issue. So your _image comparison system_ should mask out the unwanted parts in your before and after images, then do the comparison. If the mask is misaligned then there is definitely a visual change.

**Comment by SetTrend** [comment]
Actually, it's only the browser itself that's able to mask the desired parts. It cannot be the test system because it isn't aware of the DOM and what's rendered by the DOM. So, considering cross device / cross browser testing, it's the browser that's supposed to mask the corresponding DOM elements before taking the screenshot.

**Comment by jlipps** [comment]
@SetTrend couldn't the test simply request the rects of the objects in question, to pass those coordinates off to an image processing library?

**Comment by SetTrend** [comment]
This would mean a huge, unnecessary overhead, causing a lot of additional processing back and forth. In terms of implementation and processing overhead that would make a mountain out of a dust speck.

The browser is suggested to just render the elements in question as described and take a screenshot after that - that's it.

**Comment by jlipps** [comment]
If that's too much overhead, I bet a reasonably tiny JS snippet involving getBoundingClientRect, z-index, and some black divs would also be a good solution. Call it with executeScript for each element you want to hide.

I'm not one of the core spec authors but my intuition is that your request goes beyond the basic facility the spec is meant to provide. Once we provide the ability to color out divs, what about circular areas? Alpha channels? Etc. IMO the spec is designed to provide the tools to make just about any kind of automation possible, but not necessarily support direct access to those kinds of automation as part of the spec itself.

**Comment by SetTrend** [comment]
What do you believe the spec is supposed to provide? Does the spec explicitly exclude providing features for automated cross-device/cross-browser testing?

The idea you suggest would require _bespoke_ solutions to be created. That's not what a _standard_ ist about.

Don't forget the cross-device/cross-browser focus of this proposal. Devices and browsers should provide this functionality by default.

**Comment by ozcoder** [comment]
I still don't understand your point @SetTrend . *Webdriver* drives the browser and gives you a screenshot. You apply your mask and if the _DIVs_ have moved they won't be covered properly. You compare the *new masked* image with your *master masked* image. If the _DIVs_ content is visible then the comparison fails, which is what I think you want anyway. Doing it your way the new mask will be moved (because the _DIVs_ did) and the comparison will still fail, but you had to add *debug* markup.

Something like this using Imagemagick.
`compare -metric ae -fuzz 10% masked_homepage_chrome_201906031454.png ../assets/masked_master_homepage_chrome.png diff.png`

**Comment by jlipps** [comment]
@SetTrend, yes, IMO the spec explicitly excludes consideration of many features that would be useful for cross-device testing. The point of the spec isn't to handle any feature that might conceivably be useful, but to give a set of browser automation primitives which can be built upon in various ways to achieve all kinds of automation.

The solution I gave is indeed a bespoke one. That's the point. The bespoke bits should exist in the testsuite code space, or some other library, not the spec.

**Comment by SetTrend** [comment]
@jlipps : You seem to be implying that cross-device/cross-browser testing is kind of a niche, but in fact it's an essential core requirement. Testing web sites doesn't make sense if testing isn't performed across devices and browsers. So a spec missing this essential feature would be unfinished, kind of half-baked.

**Comment by jlipps** [comment]
I'll let a spec author like @andreastt or @shs96c have final say in whether what you mention is an essential feature for browser automation. IMO it might be essential for cross browser testing in general but that doesn't mean it should be part of a browser automation standard.

**Comment by andreastt** [comment]
I tend to agree with @jlipps that WebDriver seems to provide the necessary primitives to achieve what you want to do.

WebDriver is not trying to “explicitly exclude providing features for […] cross-browser testing”, as you put it, but it needs to strike a balance between a reasonably sanitised API and feature bloat. Indeed, testing is one of the primary use cases for WebDriver, and many of the APIs have convenience features built in to them. One such is the page loading strategies for navigation that otherwise would make it unreasonably difficult to detect when a document has finished loading, given the constraint of the request/response protocol.

It was suggested above that an alternative to `imageMask` would be to mask the screen captures on the local end using some combination of the elements’ bounding rects, and this is a perfectly reasonable approach. In fact, I believe this one of the techniques [Applitools](https://applitools.com/) uses in their visual AI website comparison software.

I’d like to propose an alternative that is more cost-efficient and avoids the need to post-process the screen captures. WebDriver provides a [`navigator.webdriver`](https://w3c.github.io/webdriver/#interface) DOM attribute for “co-operating user agents to inform the document that it is controlled by WebDriver, for example so that alternate code paths can be triggered during automation.”

`navigator.webdriver` is provided for _exactly_ this kind of scenario. Using that, you could imagine instrumenting your document so that elements prone to high-frequency content changes are automatically masqued.

A modest proposal:

```html
<!doctype html>
<meta charset=utf-8>
<title>Automatic masking proposal</title>

<style>
#stockprice, #change { color: green }
.mask.active {
	all: revert;
	background-color: magenta !important;
	color: magenta !important;
}
</style>

<h1>Fast Brokers</h1>

<p>Twyx Corp

<p id=stockprice class=mask>$105.75
<p id=change class=mask>+ $2.11

<script>
document.addEventListener("DOMContentLoaded", () => {
	if (navigator.webdriver) {
		for (const maskEl of document.querySelectorAll(".mask")) {
			maskEl.classList.add("active");
		}
	}
});
</script>
```

You could imagine a more technically sophisticated solution, but it serves as an example of what can be done when the document can take alternate code paths based on its runtime environment.

When a regular user visits the document, they would see this:

![Screenshot_2019-06-04 Automatic masking proposal(1)](https://user-images.githubusercontent.com/399120/58862800-672b5980-86a9-11e9-8701-115bf6cf7836.png)

But when WebDriver visits the document and `navigator.webdriver` is true, the document would be rendered this way:

![Screenshot_2019-06-04 Automatic masking proposal](https://user-images.githubusercontent.com/399120/58862808-6abee080-86a9-11e9-95ad-a5ec7e456513.png)

<hr>

That said, I’m sure the working group is happy to consider this proposal in more detail at the [forthcoming face-to-face meeting](https://www.w3.org/2019/09/TPAC/) in September. We have a track record of not adopting new features lightly and not without broader consensus and buy-in from vendors, so despite my initial apprehension it doesn’t mean this is a bad idea.

**Comment by SetTrend** [comment]
I see your point. Yet, this would require adding test code into production code.

I tend to believe that any testing features have to be kept out of production code as they potentionally enlarge the attack surface footprint. So, anything like masking images should not be in production code.
